src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReadBufferState.cs

```diff
@@ -2,32 +2,168 @@
+        private int _maxCount; // Number of bytes we need to clear before returning the buffer.
+        private bool _isFirstBlock;
+        private bool _isFinalBlock;
 
-        public ReadBufferState(int defaultBufferSize)
+        public ReadBufferState(int initialBufferSize)
         {
-            Buffer = ArrayPool<byte>.Shared.Rent(Math.Max(defaultBufferSize, JsonConstants.Utf8Bom.Length));
-            BytesInBuffer = ClearMax = 0;
-            IsFirstIteration = true;
-            IsFinalBlock = false;
+            _buffer = ArrayPool<byte>.Shared.Rent(Math.Max(initialBufferSize, JsonConstants.Utf8Bom.Length));
+            _maxCount = _count = _offset = 0;
+            _isFirstBlock = true;
+            _isFinalBlock = false;
+        }
+
+        public bool IsFinalBlock => _isFinalBlock;
+
+        public ReadOnlySpan<byte> Bytes => _buffer.AsSpan(_offset, _count);
+
+        /// <summary>
+        /// Read from the stream until either our buffer is filled or we hit EOF.
+        /// Calling ReadCore is relatively expensive, so we minimize the number of times
+        /// we need to call it.
+        /// </summary>
+        public readonly async ValueTask<ReadBufferState> ReadFromStreamAsync(
+            Stream utf8Json,
+            CancellationToken cancellationToken,
+            bool fillBuffer = true)
+        {
+            // Since mutable structs don't work well with async state machines,
+            // make all updates on a copy which is returned once complete.
+            ReadBufferState bufferState = this;
+
+            do
+            {
+                int bytesRead = await utf8Json.ReadAsync(
+#if BUILDING_INBOX_LIBRARY
+                    bufferState._buffer.AsMemory(bufferState._count),
+#else
+                    bufferState._buffer, bufferState._count, bufferState._buffer.Length - bufferState._count,
+#endif
+                    cancellationToken).ConfigureAwait(false);
+
+                if (bytesRead == 0)
+                {
+                    bufferState._isFinalBlock = true;
+                    break;
+                }
+
+                bufferState._count += bytesRead;
+            }
+            while (fillBuffer && bufferState._count < bufferState._buffer.Length);
+
+            bufferState.ProcessReadBytes();
+            return bufferState;
+        }
+
+        /// <summary>
+        /// Read from the stream until either our buffer is filled or we hit EOF.
+        /// Calling ReadCore is relatively expensive, so we minimize the number of times
+        /// we need to call it.
+        /// </summary>
+        public void ReadFromStream(Stream utf8Json)
+        {
+            do
+            {
+                int bytesRead = utf8Json.Read(
+#if BUILDING_INBOX_LIBRARY
+                    _buffer.AsSpan(_count));
+#else
+                    _buffer, _count, _buffer.Length - _count);
+#endif
+
+                if (bytesRead == 0)
+                {
+                    _isFinalBlock = true;
```

Comment:
I think this variable is mis-named. I'd rename it to something like "_endOfInput"

Reply:
(i.e. first block can also be the last one but you don't know until you're done reading)